velox/connectors/hive/iceberg/FilterUtil.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/connectors/hive/iceberg/FilterUtil.h"
+
+namespace facebook::velox::connector::hive::iceberg {
+
+using namespace facebook::velox::exec;
+using namespace facebook::velox::core;
+
+// Read 'size' values from 'valuesVector' starting at 'offset', de-duplicate
+// remove nulls and sort. Return a list of unique non-null values sorted in
+// ascending order and a boolean indicating whether there were any null values.
+template <typename T, typename U = T>
+std::pair<std::vector<T>, bool> toValues(
+    const VectorPtr& valuesVector,
+    vector_size_t offset,
+    vector_size_t size) {
+  auto simpleValues = valuesVector->as<SimpleVector<U>>();
+
+  bool nullAllowed = false;
+  std::vector<T> values;
+  values.reserve(size);
+
+  for (auto i = offset; i < offset + size; i++) {
+    if (simpleValues->isNullAt(i)) {
+      nullAllowed = true;
+    } else {
+      if constexpr (std::is_same_v<U, Timestamp>) {
+        values.emplace_back(simpleValues->valueAt(i).toMillis());
```

Comment:
@rui-mo It will create a BigIntRange to filter data values. I'm not quite sure why we need to convert the Timestamp type to bigint here; shouldn't we be using TimestampRange instead?

Reply:
@rui-mo @liujiayi771 this toValues() function is copied from velox/functions/prestosql/InPredicate.cpp. We could extract it to velox/common but there isn't a good folder yet. Any good ideas? cc @Yuhta

Reply:
@yingsu00 I see. The InPredicate is also registered by Spark sql and might needs to be fixed through providing configurable behavior. For the one in the 'hive/iceberg/FilterUtil.cpp', I assume a configuration 'kReadTimestampUnit' in the hiveConfig might help adapt to different precision used by Presto and Spark. It's fine for me to leave a TODO here and focus on the primary implementation first. Thanks.

Reply:
https://github.com/facebookincubator/velox/issues/11755

Reply:
@rui-mo Is the issue here that the precision of a Timestamp in Spark is in microseconds, and converting it to milliseconds for comparison would result in a loss of precision? For example, Timestamp(1, 999000) and Timestamp(1, 998000) would be considered the same?

Reply:
https://github.com/facebookincubator/velox/issues/11755#issuecomment-2530864206 cc: @liujiayi771